src/main/java/io/appform/nautilus/funnel/graphmanagement/ESEdgeBasedGraphBuilder.java

```diff
@@ -153,15 +154,35 @@ public Graph build(final String tenant, Context analyticsContext, GraphRequest g
 
             ArrayList<GraphNode> verticesList = new ArrayList<>(vertices.values());
             verticesList.sort((GraphNode lhs, GraphNode rhs) -> Integer.compare(lhs.getId(), rhs.getId()));
+            ImmutableList<FlatPath> paths = flatPathListBuilder.build();
+            Map<String, Integer> ranks = rankNodes(paths);
             return Graph.builder()
                     .vertices(verticesList)
                     .edges(edges)
                     .paths(flatPathListBuilder.build())
+                    .ranks(ranks)
                     .build();
-        }  catch (Exception e) {
+        } catch (Exception e) {
             log.error("Error running grouping: ", e);
             throw new NautilusException(
                     ErrorMessageTable.ErrorCode.ANALYTICS_ERROR, e);
         }
     }
+
+    private Map<String, Integer> rankNodes(List<FlatPath> paths) {
```

Comment:
I know that there are no many Tests, but can you move the algorithm to PathUtils and add a test for it?